File: python/seldon_mlmd_tools/schemas.py
import json

from pydantic import BaseModel
from typing import List, Dict, Union, Tuple, Literal

from ml_metadata.proto import metadata_store_pb2


################################ Seldon Schemas ###############################


class PydanticBase(BaseModel):
    class Config:
        validate_assignment = True


# TODO: Decide if we really need a different classes for each of the feature type


class OrdinalFeature(PydanticBase):
    name: str
    dtype: str
    qdtype: Literal["ordinal"]
    stats: Dict[str, int] = None


class CategoricalFeature(PydanticBase):
    name: str
    dtype: str
    qdtype: Literal["categorical"]
    n_categories: int
    category_map: Dict[int, str]


class OneHotCategoricalFeature(PydanticBase):
    name: str
    dtype: str
    qdtype: Literal["one-hot-categorical"]
    category_name: str
    categorical_variable_id: int


class DefaultFeature(PydanticBase):
    name: str
    dtype: str
    qdtype: str


class DataSchema(PydanticBase):
    type: str
    features: List[
        Union[
            OrdinalFeature, CategoricalFeature, OneHotCategoricalFeature, DefaultFeature
        ]
    ]


class Model(PydanticBase):
    name: str = None
    type: str = None
    task: str = None
    input_schema: DataSchema = None
    output_schema: DataSchema = None


########################## MLMD Artifact Types ################################

SELDON_GRANULAR_MODEL = "SeldonGranularModel"

MLMD_artifact_TYPES = {
    SELDON_GRANULAR_MODEL: {
        "type": metadata_store_pb2.STRING,
        "task": metadata_store_pb2.STRING,
        "input_data_schema": metadata_store_pb2.STRING,
        "output_data_schema": metadata_store_pb2.STRING,
    }
}


########################### Translation Layer #################################


def seldon_model_as_mlmd_artifact(model: Model) -> metadata_store_pb2.Artifact:
    """Convert Model schema to unsaved artifact.

    Note: both artifact's id and type_id are not set.
    """
    artifact = metadata_store_pb2.Artifact()
    artifact.name = model.name

    artifact.properties["type"].string_value = model.type
    artifact.properties["task"].string_value = model.task
    artifact.properties["input_data_schema"].string_value = model.input_schema.json()
    artifact.properties["output_data_schema"].string_value = model.output_schema.json()

    return artifact


def mlmd_artifact_as_seldon_model(artifact: metadata_store_pb2.Artifact) -> Model:
    model = Model(
        name=artifact.name,
        task=artifact.properties["task"].string_value,
        type=artifact.properties["type"].string_value,
        input_schema=json.loads(artifact.properties["input_data_schema"].string_value),
        output_schema=json.loads(
            artifact.properties["output_data_schema"].string_value
        ),
    )
    return model

File: golang/schemas/schemas.go
package schemas

import (
	"encoding/json"
	pb "github.com/seldonio/seldon-mlmd-tools/golang/proto/ml_metadata"
)

type FeatureMetadata interface {
	isFeatureMetadata()
}

type DefaultFeature struct {
	Name   string `json:"name"`
	DType  string `json:"dtype"`
	QDType string `json:"qdtype"`
}

func (*DefaultFeature) isFeatureMetadata() {}

type OrdinalFeature struct {
	Name   string         `json:"name"`
	DType  string         `json:"dtype"`
	QDType string         `json:"qdtype"`
	Stats  map[string]int `json:"stats"`
}

func (*OrdinalFeature) isFeatureMetadata() {}

type CategoricalFeature struct {
	Name        string         `json:"name"`
	DType       string         `json:"dtype"`
	QDType      string         `json:"qdtype"`
	NCategories int            `json:"n_categories"`
	CategoryMap map[int]string `json:"category_map"`
}

func (*CategoricalFeature) isFeatureMetadata() {}

type OneHotCategoricalFeature struct {
	Name               string `json:"name"`
	DType              string `json:"dtype"`
	QDType             string `json:"qdtype"`
	CategoryName       string `json:"category_name"`
	CategoryVariableId int    `json:"categorical_variable_id"`
}

func (*OneHotCategoricalFeature) isFeatureMetadata() {}

type DataSchema struct {
	Type     string            `json:"type"`
	Features []FeatureMetadata `json:"features"`
}

type RawDataSchema struct {
	Type     string            `json:"type"`
	Features []json.RawMessage `json:"features"`
}

func NewDataSchemaFromValue(value *pb.Value) *DataSchema {
	var rawDataSchema RawDataSchema
	json.Unmarshal([]byte(value.GetStringValue()), &rawDataSchema)

	features := make([]FeatureMetadata, 0, len(rawDataSchema.Features))
	for _, rawFeatureJson := range rawDataSchema.Features {

		var rawFeature map[string]interface{}
		json.Unmarshal([]byte(rawFeatureJson), &rawFeature)

		switch featureType := rawFeature["qdtype"].(string); featureType {
		case "ordinal":
			var feature OrdinalFeature
			json.Unmarshal([]byte(rawFeatureJson), &feature)
			features = append(features, &feature)
		case "categorical":
			var feature CategoricalFeature
			json.Unmarshal([]byte(rawFeatureJson), &feature)
			features = append(features, &feature)
		case "one-hot-categorical":
			var feature OneHotCategoricalFeature
			json.Unmarshal([]byte(rawFeatureJson), &feature)
			features = append(features, &feature)
		default:
			var feature DefaultFeature
			json.Unmarshal([]byte(rawFeatureJson), &feature)
			features = append(features, &feature)
		}

	}

	return &DataSchema{
		Type:     rawDataSchema.Type,
		Features: features,
	}
}

type GranularModelMetadata struct {
	ModelName        string      `json:"name"`
	ModelType        string      `json:"type"`
	ModelTask        string      `json:"task"`
	InputDataSchema  *DataSchema `json:"input_data_schema"`
	OutputDataSchema *DataSchema `json:"output_data_schema"`
}

func NewGranularModelMetadataFromArtifact(artifact *pb.Artifact) *GranularModelMetadata {
	properties := artifact.GetProperties()
	metadata := GranularModelMetadata{
		ModelName:        artifact.GetName(),
		ModelType:        properties["type"].GetStringValue(),
		ModelTask:        properties["task"].GetStringValue(),
		InputDataSchema:  NewDataSchemaFromValue(properties["input_data_schema"]),
		OutputDataSchema: NewDataSchemaFromValue(properties["output_data_schema"]),
	}

	return &metadata
}
